clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/fragments/ProtocolFragment.kt

```diff
@@ -1,43 +1,420 @@
+                    logger?.update(data.size)
+
+                    if (data.size < dataLen) {
+                        gatt.readCharacteristic(characteristic)
+                    } else {
+                        val msg = data
+                        dataLen = 0
+                        data = byteArrayOf()
+                        it.onMessageRead(msg)
+                    }
+                }
+            }
+        }
+
+        /*
+            As messages the client sends will never be bigger than the MTU, we don't need to care
+            about chunking them.
+         */
+        override fun onCharacteristicWrite(gatt: BluetoothGatt?, characteristic: BluetoothGattCharacteristic?, status: Int) {
+            super.onCharacteristicWrite(gatt, characteristic, status)
+            protocol?.let { proto ->
+                characteristic?.let {
+                    proto.onMessageWrite()
+                }
+            }
+        }
+    }
+
+    /*
+        Fragment lifecycle methods:
+     */
 
     override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-        setHasOptionsMenu(true)
-        device = requireArguments().getSerializable("device") as Device?
-        return inflater.inflate(R.layout.fragment_device_list, container, false)
+        val menuHost = requireActivity()
+        menuHost.addMenuProvider(object: MenuProvider {
+            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
+                menuInflater.inflate(R.menu.action_bar, menu)
+            }
+            override fun onPrepareMenu(menu: Menu) {
+                super.onPrepareMenu(menu)
+                menu.findItem(R.id.action_edit).isVisible = false
+                menu.findItem(R.id.action_add).isVisible = false
+            }
+            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
+                return false
+            }
+        })
+        (activity as MainActivity).showUpButton()
+        return inflater.inflate(R.layout.fragment_protocol, container, false)
     }
 
+    @SuppressLint("MissingPermission")
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
-        (requireActivity() as MainActivity).supportActionBar?.title = if (device != null) {
-            "Attestation in progress"
+		val device = requireArguments().getSerializable("device") as Device
+        if (device.name.isEmpty()) {
+            state = State.ENROLLMENT
+            activity?.title = "Enrollment in progress"
+        } else {
+            state = State.AUTHENTICATION
+            activity?.title = "Attestation in progress"
+        }
+        logger = Logger(activity as MainActivity?, view.findViewById(R.id.logger_text_view), onError = {
+            // TODO: Ask the user if he wants to inspect the logs or go back to the menu
+        })
+
+        if (!hasBluetoothPermission()) {
```

Comment:
Do this only later, BLUETOOTH_CONNECT is a runtime permission on API 31 and above.
https://developer.android.com/guide/topics/connectivity/bluetooth/permissions